File: cxx-qt-gen/test_outputs/types_primitive_property.rs
#[cxx::bridge(namespace = "cxx_qt::my_object")]
mod ffi {
    unsafe extern "C++" {
        include!("cxx-qt-gen/include/my_object.cxxqt.h");

        #[cxx_name = "MyObject"]
        type MyObjectQt;

        #[rust_name = "boolean"]
        fn getBoolean(self: &MyObjectQt) -> bool;
        #[rust_name = "set_boolean"]
        fn setBoolean(self: Pin<&mut MyObjectQt>, value: bool);

        #[rust_name = "float_32"]
        fn getFloat32(self: &MyObjectQt) -> f32;
        #[rust_name = "set_float_32"]
        fn setFloat32(self: Pin<&mut MyObjectQt>, value: f32);

        #[rust_name = "float_64"]
        fn getFloat64(self: &MyObjectQt) -> f64;
        #[rust_name = "set_float_64"]
        fn setFloat64(self: Pin<&mut MyObjectQt>, value: f64);

        #[rust_name = "int_8"]
        fn getInt8(self: &MyObjectQt) -> i8;
        #[rust_name = "set_int_8"]
        fn setInt8(self: Pin<&mut MyObjectQt>, value: i8);

        #[rust_name = "int_16"]
        fn getInt16(self: &MyObjectQt) -> i16;
        #[rust_name = "set_int_16"]
        fn setInt16(self: Pin<&mut MyObjectQt>, value: i16);

        #[rust_name = "int_32"]
        fn getInt32(self: &MyObjectQt) -> i32;
        #[rust_name = "set_int_32"]
        fn setInt32(self: Pin<&mut MyObjectQt>, value: i32);

        #[rust_name = "uint_8"]
        fn getUint8(self: &MyObjectQt) -> u8;
        #[rust_name = "set_uint_8"]
        fn setUint8(self: Pin<&mut MyObjectQt>, value: u8);

        #[rust_name = "uint_16"]
        fn getUint16(self: &MyObjectQt) -> u16;
        #[rust_name = "set_uint_16"]
        fn setUint16(self: Pin<&mut MyObjectQt>, value: u16);

        #[rust_name = "uint_32"]
        fn getUint32(self: &MyObjectQt) -> u32;
        #[rust_name = "set_uint_32"]
        fn setUint32(self: Pin<&mut MyObjectQt>, value: u32);
    }

    extern "Rust" {
        #[cxx_name = "MyObjectRust"]
        type MyObject;
    }

    unsafe extern "C++" {
        include ! (< QtCore / QObject >);
        include!("cxx-qt-lib/include/convert.h");
        include!("cxx-qt-lib/include/cxxqt_thread.h");

        type MyObjectCxxQtThread;

        #[cxx_name = "unsafeRust"]
        fn rust(self: &MyObjectQt) -> &MyObject;

        #[cxx_name = "qtThread"]
        fn qt_thread(self: &MyObjectQt) -> UniquePtr<MyObjectCxxQtThread>;
        fn queue(self: &MyObjectCxxQtThread, func: fn(ctx: Pin<&mut MyObjectQt>)) -> Result<()>;

        #[rust_name = "new_cpp_object"]
        #[namespace = "cxx_qt::my_object::cxx_qt_my_object"]
        fn newCppObject() -> UniquePtr<MyObjectQt>;
    }

    extern "C++" {
        #[cxx_name = "unsafeRustMut"]
        unsafe fn rust_mut(self: Pin<&mut MyObjectQt>) -> Pin<&mut MyObject>;
    }

    extern "Rust" {
        #[cxx_name = "createRs"]
        #[namespace = "cxx_qt::my_object::cxx_qt_my_object"]
        fn create_rs() -> Box<MyObject>;

        #[cxx_name = "initialiseCpp"]
        #[namespace = "cxx_qt::my_object::cxx_qt_my_object"]
        fn initialise_cpp(cpp: Pin<&mut MyObjectQt>);
    }
}

pub use self::cxx_qt_ffi::*;
mod cxx_qt_ffi {
    use super::ffi::*;

    pub type FFICppObj = super::ffi::MyObjectQt;
    type UniquePtr<T> = cxx::UniquePtr<T>;

    unsafe impl Send for MyObjectCxxQtThread {}

    use std::pin::Pin;

    #[derive(Default)]
    pub struct MyObject;

    impl MyObject {}

    impl MyObjectQt {
        pub fn grab_values_from_data(mut self: Pin<&mut Self>, mut data: Data) {
            self.as_mut().set_boolean(data.boolean);
            self.as_mut().set_float_32(data.float_32);
            self.as_mut().set_float_64(data.float_64);
            self.as_mut().set_int_8(data.int_8);
            self.as_mut().set_int_16(data.int_16);
            self.as_mut().set_int_32(data.int_32);
            self.as_mut().set_uint_8(data.uint_8);
            self.as_mut().set_uint_16(data.uint_16);
            self.as_mut().set_uint_32(data.uint_32);
        }
    }

    #[derive(Default)]
    pub struct Data {
        boolean: bool,
        float_32: f32,
        float_64: f64,
        int_8: i8,
        int_16: i16,
        int_32: i32,
        uint_8: u8,
        uint_16: u16,
        uint_32: u32,
    }

    impl From<&MyObjectQt> for Data {
        fn from(value: &MyObjectQt) -> Self {
            Self {
                boolean: value.boolean().into(),
                float_32: value.float_32().into(),
                float_64: value.float_64().into(),
                int_8: value.int_8().into(),
                int_16: value.int_16().into(),
                int_32: value.int_32().into(),
                uint_8: value.uint_8().into(),
                uint_16: value.uint_16().into(),
                uint_32: value.uint_32().into(),
            }
        }
    }

    pub fn create_rs() -> std::boxed::Box<MyObject> {
        std::default::Default::default()
    }

    pub fn initialise_cpp(cpp: std::pin::Pin<&mut MyObjectQt>) {
        cpp.grab_values_from_data(Data::default());
    }
}

File: cxx-qt-gen/test_outputs/types_primitive_property.cpp
#include "cxx-qt-gen/include/my_object.cxxqt.h"

namespace cxx_qt::my_object {

MyObject::MyObject(QObject* parent)
  : QObject(parent)
  , m_rustObj(cxx_qt::my_object::cxx_qt_my_object::createRs())
  , m_rustObjMutex(std::make_shared<std::mutex>())
  , m_cxxQtThreadObj(
      std::make_shared<rust::cxxqtlib1::CxxQtGuardedPointer<MyObject>>(this))
{
}

MyObject::~MyObject()
{
  const auto guard = std::unique_lock(m_cxxQtThreadObj->mutex);
  m_cxxQtThreadObj->ptr = nullptr;
}

const MyObjectRust&
MyObject::unsafeRust() const
{
  return *m_rustObj;
}

MyObjectRust&
MyObject::unsafeRustMut()
{
  return *m_rustObj;
}

std::unique_ptr<MyObjectCxxQtThread>
MyObject::qtThread() const
{
  return std::make_unique<MyObjectCxxQtThread>(m_cxxQtThreadObj,
                                               m_rustObjMutex);
}

bool
MyObject::getBoolean() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<bool, bool>{}(
    m_rustObj->getBoolean(*this));
}

void
MyObject::setBoolean(bool value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setBoolean(*this, value);
}

void
MyObject::emitBooleanChanged()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "booleanChanged", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

float
MyObject::getFloat32() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<float, float>{}(
    m_rustObj->getFloat32(*this));
}

void
MyObject::setFloat32(float value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setFloat32(*this, value);
}

void
MyObject::emitFloat32Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "float32Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

double
MyObject::getFloat64() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<double, double>{}(
    m_rustObj->getFloat64(*this));
}

void
MyObject::setFloat64(double value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setFloat64(*this, value);
}

void
MyObject::emitFloat64Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "float64Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

qint8
MyObject::getInt8() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<qint8, qint8>{}(
    m_rustObj->getInt8(*this));
}

void
MyObject::setInt8(qint8 value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setInt8(*this, value);
}

void
MyObject::emitInt8Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "int8Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

qint16
MyObject::getInt16() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<qint16, qint16>{}(
    m_rustObj->getInt16(*this));
}

void
MyObject::setInt16(qint16 value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setInt16(*this, value);
}

void
MyObject::emitInt16Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "int16Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

qint32
MyObject::getInt32() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<qint32, qint32>{}(
    m_rustObj->getInt32(*this));
}

void
MyObject::setInt32(qint32 value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setInt32(*this, value);
}

void
MyObject::emitInt32Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "int32Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

quint8
MyObject::getUint8() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<quint8, quint8>{}(
    m_rustObj->getUint8(*this));
}

void
MyObject::setUint8(quint8 value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setUint8(*this, value);
}

void
MyObject::emitUint8Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "uint8Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

quint16
MyObject::getUint16() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<quint16, quint16>{}(
    m_rustObj->getUint16(*this));
}

void
MyObject::setUint16(quint16 value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setUint16(*this, value);
}

void
MyObject::emitUint16Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "uint16Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

quint32
MyObject::getUint32() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<quint32, quint32>{}(
    m_rustObj->getUint32(*this));
}

void
MyObject::setUint32(quint32 value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setUint32(*this, value);
}

void
MyObject::emitUint32Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "uint32Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

} // namespace cxx_qt::my_object

namespace cxx_qt::my_object::cxx_qt_my_object {
std::unique_ptr<MyObject>
newCppObject()
{
  return std::make_unique<MyObject>();
}
} // namespace cxx_qt::my_object::cxx_qt_my_object
